generate/templates/manual/remote/ls.cc

```diff
@@ -0,0 +1,136 @@
+NAN_METHOD(GitRemote::ReferenceList)
+{
+  if (info.Length() == 0 || !info[0]->IsObject()) {
+    return Nan::ThrowError("RemoteCallbacks callbacks is required.");
+  }
+  if (info.Length() == 1 || !info[1]->IsObject()) {
+    return Nan::ThrowError("ProxyOptions proxy_opts is required.");
+  }
+
+  if (info.Length() == 2 || !(Nan::To<bool>(info[2]).FromJust())) {
+    return Nan::ThrowError("Array, String Object, or string custom_headers is required.");
+  }
+  if (info.Length() == 3 || !info[3]->IsFunction()) {
+    return Nan::ThrowError("Callback is required and must be a Function.");
+  }
+
+  ReferenceListBaton* baton = new ReferenceListBaton;
+
+  baton->error_code = GIT_OK;
+  baton->error = NULL;
+  baton->out = new std::vector<git_remote_head*>;
+  baton->remote = Nan::ObjectWrap::Unwrap<GitRemote>(info.This())->GetValue();
+  baton->callbacks = Nan::ObjectWrap::Unwrap<GitRemoteCallbacks>(info[0]->ToObject())->GetValue();
+  baton->proxy_opts = Nan::ObjectWrap::Unwrap<GitProxyOptions>(info[1]->ToObject())->GetValue();
+  baton->custom_headers = StrArrayConverter::Convert(info[2]);
+
+  Nan::Callback *callback = new Nan::Callback(Local<Function>::Cast(info[3]));
+  ReferenceListWorker *worker = new ReferenceListWorker(baton, callback);
+  worker->SaveToPersistent("remote", info.This());
+  if (!info[1]->IsUndefined() && !info[0]->IsNull())
+    worker->SaveToPersistent("callbacks", info[0]->ToObject());
+  if (!info[2]->IsUndefined() && !info[1]->IsNull())
+    worker->SaveToPersistent("proxy_opts", info[1]->ToObject());
+  if (!info[3]->IsUndefined() && !info[2]->IsNull())
+    worker->SaveToPersistent("custom_headers", info[2]->ToObject());
+  Nan::AsyncQueueWorker(worker);
+  return;
+}
+
+void GitRemote::ReferenceListWorker::Execute()
+{
+  giterr_clear();
+
+  {
+    LockMaster lockMaster(
+      /*asyncAction: */true,
+      baton->remote,
+      baton->callbacks,
+      baton->proxy_opts,
+      baton->custom_headers
+    );
+    baton->error_code = git_remote_connect(
```

Comment:
Instead of connecting (and disconnecting) as part of `ReferenceList`, it could be just a wrapper over `git_remote_ls`, and that way the user can use it with an already connected remote.